text/3305-static-in-pattern.md

```diff
@@ -0,0 +1,285 @@
+```rust
+static TRUE: bool = true;
+static FALSE: bool = false;
+
+fn foo(scrutinee: bool) {
+    match scrutinee {
+        TRUE | FALSE => println!("bar"),
+
+        // The compiler will throw an error if you remove this branch;
+        // it is not allowed to look into the values of the statics
+        // to determine that it is unreachable.
+        _ => println!("baz"),
+    }
+}
+```
+
+As an exception, when all safe values of a type are structurally equal, the compiler is allowed to see that the match will always succeed.
+
+```rust
+// Not all `&()` are bitwise equal,
+// but they are structurally equal,
+// that is what matters.
+static ONE_TRUE_VALUE: &() = &();
+
+fn foo(scrutinee: &()) {
+    match scrutinee {
+        ONE_TRUE_VALUE => println!("only one branch"),
+        // No need for a wildcard.
+        // The above match always succeeds.
+    }
+}
+```
+
+Visibility and `#[non_exhaustive]` can affect whether the compiler can tell that all values of the type are structurally equal.
+
+```rust
+mod stuff {
+    #[derive(PartialEq, Eq)]
+    pub(super) struct PrivateZst(());
+
+    pub(super) static PRIVATE_ZST: PrivateZst = PrivateZst(());
+}
+
+fn foo(scrutinee: stuff::PrivateZst) {
+    match scrutinee {
+        stuff::PRIVATE_ZST => println!("secrets abound"),
+        // `stuff::PrivateZst` has a field that's not visible in this scope,
+        // so we can't tell that all values are equivalent.
+        // The wildcard branch is required.
+        _ => println!("incorrect password"),
+    }
+}
+```
+
+```rust
+// crate `stuff`
+#[derive(PartialEq, Eq)]
+#[non_exhaustive]
+pub struct PrivateZst();
+
+pub static PRIVATE_ZST: PrivateZst = PrivateZst();
+
+// main crate
+extern crate stuff;
+
+fn foo(scrutinee: stuff::PrivateZst) {
+    match scrutinee {
+        stuff::PRIVATE_ZST => println!("secrets abound"),
+        // `stuff::PrivateZst` is marked `#[non_exhaustive]`
+        // and comes from an external crate,
+        // so we can't tell that all values are equivalent.
+        // The wildcard branch is required.
+        _ => println!("incorrect password"),
+    }
+}
+```
+
+Static patterns can be nested in other patterns:
+
+```rust
+static ONE: i32 = 1;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        ONE | 2 => println!("a"),
+        _ => (),
+    }
+
+    match (scrutinee, scrutinee) {
+        (ONE, ONE) =>  println!("a"),
+        _ => (),
+    }
+}
+```
+
+The examples above all use `match`, but statics would be allowed in all other language constructs that use patterns, including `let`, `if let`, and function parameters. However, as statics cannot be used in const contexts, static patterns are be unavailable there as well.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+This change slightly weakens the rule that patterns can only rely on compile-time information. In addition, static patterns may have slightly worse performance than the equivalent constant patterns.
+
+# Rationale and alternatives
```

Comment:
an unmentioned alternative is to supply an entirely new kind of pattern that can read from path expressions. So something like

```rust
match foo {
    { FOO } => {},
    { x.field } => {},
}
```

which is effectively sugar for

```rust
match foo {
    pat if pat == FOO => {},
    pat if pat == x.field => {},
}
```

and can be desugared at HIR lowering time without requiring any adjustment to pattern matching

Reply:
> path expressions

`x.field` isn't a *path* expression, did you mean "place"? Also, what's the specific goal here with that restriction, avoiding expressions that would compute a new value on the fly, and having to execute those *in the middle of* pattern-matching? Or am I missing a simpler reason?

> can be desugared at HIR lowering time

This could not limit e.g. `x.y` ending up as `Deref::deref(&x).y` (assuming you mean AST->HIR lowering).
But also I see no reason to limit this in any way more than match guards.

<hr/>

I would try to optimize for the "is equal to variable" usecase, in that case.
And for that you *really* have to make it distinct from `x => ...` style patterns - so I don't think the braces are enough.

I thought I'd seen one or more languages that had such a thing but I couldn't find anything on my own so I asked @solson and they remembered [that Magpie has them](https://magpie-lang.org/patterns.html#equality-patterns) (not a language I'm familiar with though).
<sub>(Another thing that came up was that there's also a pattern feature in Agda and Lean that I'm pretty sure in the past I misread as such a feature - but no, Agda/Lean's "dot patterns"/"inaccessible terms" are *statically* redundant and don't actually do any comparisons, they're almost "type-checked documentation" that some value *ends up forced* to be equal to another value - can't think of any plausible Rust examples, likely only make sense with dependent typing)</sub>

Using that syntax, we'd end up with:
```rust
match foo {
    (== FOO) => {},
    (== x.field) => {},
}
```
(thought there's probably a bunch of variations based on whether you place the parens around `==` or *after* it, and whether you use `()` or `{}` etc. etc.)

What strikes me about that is that if we consider generalized guards of the form:
```rust
match (foo, bar) {
    (a if a == FOO, b if b == BAR) => ...
    _ => ...
}
```
Then you can start to see a resemblance. The main thing we're doing is removing the name.
So `(== FOO)` is kind of like if `if _ == FOO` was a valid pattern (in a similar vein to the idea of letting `_ == FOO` be a shorthand for `|x| x == FOO`)

At the end of the day, the bulk of the usefulness of such sugar for `==` is *purely aesthetic*: there is no way the compiler could know *the value* of what you're equating to, so there's no additional pattern reasoning, it's just shorter syntax for a guard.

It might be worth it but I would suggest opening a separate RFC, as opposed to considering it as a viable alternative to this one.

<hr/>

To be more precise, the stated motivation of this RFC appears to be allowing `libc` to replace constants with `extern { static }` but that is not really portable - I'm surprised the C equivalent doesn't just break existing software, e.g. something like this ([check out on godbolt](https://godbolt.org/z/6TPEqqKzq)):
```c
#include <errno.h>

enum {
    ErrInvalid = EINVAL,
    ErrUnimplemented = ENOSYS,
    ErrMissingFileOrDir = ENOENT,
    // ...
};
```

Only making it "work" with *some* syntax isn't the same as it actually *working everywhere*.
(You can easily replicate the above C example into Rust, or use a constant in an array length, both of which require real constants and cannot be substituted by some initialized-before-main variable)

Reply:
> (Another thing that came up was that there's also a pattern feature in Agda and Lean that I'm pretty sure in the past I misread as such a feature - but no, Agda/Lean's "dot patterns"/"inaccessible terms" are statically redundant and don't actually do any comparisons, they're almost "type-checked documentation" that some value ends up forced to be equal to another value - can't think of any plausible Rust examples, likely only make sense with dependent typing)

These correspond basically to ['never patterns'](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/) in Rust, though with dependent types they are a lot more complicated.

Reply:
> These correspond basically to ['never patterns'](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/) in Rust, though with dependent types they are a lot more complicated.

I'm pretty sure the two are different? (I'm talking about [dot patterns](https://agda.readthedocs.io/en/latest/language/function-definitions.html#dot-patterns) while "never patterns" sound like [absurd patterns](https://agda.readthedocs.io/en/latest/language/function-definitions.html#absurd-patterns))

Reply:
(ah, sorry, I misremembered)

Reply:
> I'm surprised the C equivalent doesn't just break existing software (as [POSIX requires `#define` macros](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/errno.h.html#tag_13_09_03)), e.g. something like this ([check out on godbolt](https://godbolt.org/z/6TPEqqKzq)):

Or simply `switch`ing on `errno`.  C requires `switch` alternatives to be constant expressions and variables are not constant expressions (even if they are `const` global variables).  Sure enough, [this is an issue with cosmopolitan-libc](https://godbolt.org/z/6TPEqqKzq).